[IntuneDeviceFeaturesConfigurationPolicyIOS] New module

dannyKBjj opened the pull request:
#### Pull Request (PR) description
Adds support for Devices>Configuration>iOS>Templates>Device Features (#microsoft.graph.iosDeviceFeaturesConfiguration)

#### This Pull Request (PR) fixes the following issues
Initial release

#### Task list

<!--
    To aid community reviewers in reviewing and merging your PR, please take the time to run
    through the below checklist and make sure your PR has everything updated as required.

    Change to [x] for each task in the task list that applies to your PR. For those task that
    don't apply to you PR, leave those as is.
-->

- [x ] Added an entry to the change log under the Unreleased section of the file CHANGELOG.md.
      Entry should say what was changed and how that affects users (if applicable), and
      reference the issue being resolved (if applicable).
- [x ] Resource parameter descriptions added/updated in the schema.mof.
- [x ] Resource documentation added/updated in README.md.
- [ x] Resource settings.json file contains all required permissions.
- [x ] Examples appropriately added/updated.
- [x ] Unit tests added/updated.
- [x ] New/changed code adheres to [DSC Community Style Guidelines](https://dsccommunity.org/styleguidelines).

dannyKBjj committed:
Working correctly, need unit tests

dannyKBjj committed:
New module - IntuneDeviceFeaturesConfigurationPolicyIOS

dannyKBjj committed:
Fixed example mof files

dannyKBjj committed:
Removed comments from schema.mof

dannyKBjj commented:
seems to be moaning about my schema.mof, but there's literally nothing wrong with it so far as I can see. Resource works fine, mofs compile and all objects defined. It had some comments in it and I thought they might have been tripping things up, but have removed the comments and it still fails...

dannyKBjj commented:
would appreciate some help on this one!

dannyKBjj committed:
added missing class defs

Unsure what these properties map to in GUI, if anything? couldn't figure out how to populate them

dannyKBjj committed:
removed comment from test script

dannyKBjj committed:
corrected verbose messages

FabienTschanz commented on Modules/Microsoft365DSC/Examples/Resources/IntuneDeviceFeaturesConfigurationPolicyIOS/1-Create.ps1:
Please make sure that this file is UTF-8 encoded.

dannyKBjj replied:
done

FabienTschanz commented on Modules/Microsoft365DSC/Examples/Resources/IntuneDeviceFeaturesConfigurationPolicyIOS/2-Update.ps1:
Please make sure that this file is UTF-8 encoded.

dannyKBjj replied:
done

FabienTschanz commented on Modules/Microsoft365DSC/Examples/Resources/IntuneDeviceFeaturesConfigurationPolicyIOS/3-Remove.ps1:
Please make sure that this file is UTF-8 encoded.

dannyKBjj replied:
done

FabienTschanz commented on Modules/Microsoft365DSC/DSCResources/MSFT_IntuneDeviceFeaturesConfigurationPolicyIOS/MSFT_IntuneDeviceFeaturesConfigurationPolicyIOS.psm1:
```diff
@@ -0,0 +1,1202 @@
+        $AccessTokens
+
+    )
+
```
Please add a line that states something like the following: `Write-Verbose -Message "Getting configuration of the Intune Device Features Configuration Policy for iOS with Id {$Id} and DisplayName {$DisplayName}"`

dannyKBjj replied:
added

FabienTschanz commented on Modules/Microsoft365DSC/DSCResources/MSFT_IntuneDeviceFeaturesConfigurationPolicyIOS/MSFT_IntuneDeviceFeaturesConfigurationPolicyIOS.psm1:
```diff
@@ -0,0 +1,1202 @@
+            $nullResult = $PSBoundParameters
+            $nullResult.Ensure = 'Absent'
+
+            if (-not [string]::IsNullOrWhiteSpace($id))
```
Can you make sure that the properties are properly cased? E.g. `$id` (lowercase) should actually be uppercase `$Id`.

dannyKBjj replied:
replaced all instances of $id with $Id

FabienTschanz commented on Modules/Microsoft365DSC/DSCResources/MSFT_IntuneDeviceFeaturesConfigurationPolicyIOS/MSFT_IntuneDeviceFeaturesConfigurationPolicyIOS.psm1:
```diff
@@ -0,0 +1,1202 @@
+
+    if ($Ensure -eq 'Present' -and $currentInstance.Ensure -eq 'Absent')
+    {
+        Write-Verbose -Message "Creating {$DisplayName}"
```
Please update the verbose messages to more descriptive ones, e.g. `Creating an Intune Device Features Configuration Policy for iOS with DisplayName {XYZ}`

dannyKBjj replied:
done

FabienTschanz commented on Modules/Microsoft365DSC/DSCResources/MSFT_IntuneDeviceFeaturesConfigurationPolicyIOS/MSFT_IntuneDeviceFeaturesConfigurationPolicyIOS.psm1:
```diff
@@ -0,0 +1,1202 @@
+            
+            if ($Results.HomeScreenDockIcons)
+            {
+                $currentDSCBlock = Convert-DSCStringParamToVariable -DSCBlock $currentDSCBlock -ParameterName "HomeScreenDockIcons" -isCIMArray:$True
+            }
+
+            if ($Results.HomeScreenPages)
+            {
+                $currentDSCBlock = Convert-DSCStringParamToVariable -DSCBlock $currentDSCBlock -ParameterName "HomeScreenPages" -isCIMArray:$True
+            }
+
+            if ($Results.WallpaperImage)
+            {
+                $currentDSCBlock = Convert-DSCStringParamToVariable -DSCBlock $currentDSCBlock -ParameterName "WallpaperImage" -isCIMArray:$True
+            }
+
+            if ($Results.IosSingleSignOnExtension)
+            {
+                $currentDSCBlock = Convert-DSCStringParamToVariable -DSCBlock $currentDSCBlock -ParameterName "IosSingleSignOnExtension" -isCIMArray:$True
+            }
+
+            if ($Results.NotificationSettings)
+            {
+                $currentDSCBlock = Convert-DSCStringParamToVariable -DSCBlock $currentDSCBlock -ParameterName "NotificationSettings" -isCIMArray:$True
+            }
+
+            if ($Results.SingleSignOnSettings)
+            {
+                $currentDSCBlock = Convert-DSCStringParamToVariable -DSCBlock $currentDSCBlock -ParameterName "SingleSignOnSettings" -isCIMArray:$True
+            }
+
+            if ($Results.Assignments)
+            {
+                $isCIMArray = $false
+                if ($Results.Assignments.getType().Fullname -like '*[[\]]')
+                {
+                    $isCIMArray = $true
+                }
+                $currentDSCBlock = Convert-DSCStringParamToVariable -DSCBlock $currentDSCBlock -ParameterName 'Assignments' -IsCIMArray:$isCIMArray
+            }
```
You don't need those anymore (just as `Update-M365DSCExportAuthenticationResults`), replace them by adding `-NoEscape @('<place all the complex property names like Assignments, SingleSignOnSettings etc. here')` after `-Credential $Credential` from `Get-M365DSCExportContentForResource`.

dannyKBjj replied:
Hi, Sorry Fabien I don't quite understand this?

You want me to do the following?

Delete all the lines highlighted.

Replace with 

Update-M365DSCExportAuthenticationResults -NoEscape @('AirPrintDestinations', 'ContentFilterSettings','HomeScreenDockIcons','HomeScreenPages','WallpaperImage',IosSingleSignOnExtension','NotificationSettings','SingleSignOnSettings','Assignments')

I can't find any other module using that cmdlet and can't find anything about with Google?

FabienTschanz replied:
No, you can remove the `Update-M365DSCExportAuthenticationResults` command as well. Add the `-NoEscape` parameter to `Get-M365DSCExportContentForResource` after the `-Credential $Credential`. Take a look at the other resources, e.g. `IntuneDeviceRemediation`.

dannyKBjj replied:
Sorry if I'm being an idiot, but I looked at IntuneDeviceRemediation and it has:
            $Results = Update-M365DSCExportAuthenticationResults -ConnectionMode $ConnectionMode `
                -Results $Results
on line 833 and doesn't use the -NoEscape parameter when calling Get-M365DSCExportContentForResource on line 891?

So far as I can see, it is basically the same as mine!

dannyKBjj replied:
[Export-TargetResource.txt](https://github.com/user-attachments/files/18823854/Export-TargetResource.txt)

You want me to do this?

FabienTschanz replied:
@dannyKBjj Exactly, yes. After `-Credential $Credential`, you missed a backtick there for a new line. Correct is: 

```powershell
$currentDSCBlock = Get-M365DSCExportContentForResource -ResourceName $ResourceName `
    -ConnectionMode $ConnectionMode `
    -ModulePath $PSScriptRoot `
    -Results $Results `
    -Credential $Credential ` # <-- Put a backtick here
    -NoEscape @('AirPrintDestinations', 'ContentFilterSettings','HomeScreenDockIcons','HomeScreenPages','WallpaperImage','IosSingleSignOnExtension','NotificationSettings','SingleSignOnSettings','Assignments')
```

Rest is perfect 👍

FabienTschanz commented on Modules/Microsoft365DSC/DSCResources/MSFT_IntuneDeviceFeaturesConfigurationPolicyIOS/MSFT_IntuneDeviceFeaturesConfigurationPolicyIOS.psm1:
```diff
@@ -0,0 +1,1202 @@
+            CertificateThumbprint    = $CertificateThumbprint
+            Managedidentity          = $ManagedIdentity.IsPresent
+            AccessTokens             = $AccessTokens
+            AirPrintDestinations     = Convert-ComplexObjectToHashtableArray $getValue.AdditionalProperties.airPrintDestinations
```
We normally use the DRG to create a conversion directly in the resource, just like it's in the `IntuneDeviceRemediation` resource without custom functions that are specific to a resource.

dannyKBjj replied:
I looked at the resource you mentioned. I deliberately avoided that approach because I thought it made the code hard to read and I had *a lot* of complex types to convert - I just felt it would have resulted in a mess. So I wrote a function to do it.

Ironically I considered doing something similar for the export- where you mention using Update-M365DSCExportAuthenticationResults (which I need to check how to use exactly, see my other comment). That section gives an idea of what a mess Get-TargetResource would be if I typed each structure out by hand!

To be honest I'd rather not type out all the hash tables - it'd be quite a job and I've a working solution that I thought was actually easier to read and maintain?

Any chance we can just leave this one 'as is' ?

FabienTschanz replied:
I do understand why you did it that way. What do you think @NikCharlebois?

dannyKBjj replied:
Thinking about it, the other issue is that the object structure can differ depending on the oDataType for a couple of these as well, which is another reason I feel that dynamically building the hashtables is the way forward here...

dannyKBjj replied:
@NikCharlebois ?

ykuijs replied:
@ricmestre, as the person that is doing a lot with Intune, what is your opinion on this item?

dannyKBjj committed:
latest changes

Includes changes in all resolved conversations

dannyKBjj committed:
Requested changes made to Export-DscResource

also added processing for SingleSignOnSettings as noticed an issue

dannyKBjj committed:
Using -Noescape parameter. Removed $results.Assignment block.

dannyKBjj committed:
Code tweaks

Unsure if something has changed, as I swear I tested this previously, but since upgrading to latest version my module wasn't treating certain CIM values as string, even though that's how the mof defined it, thus causing Test- to fail... either way, I spotted the problem and fixed.
Devices>Configuration>iOS>Device Features>Single Sign-on app extension>Additional Configuration now correctly handles boolean, string and integer values from the same class definition.

ykuijs committed:
Update CHANGELOG with unreleased changes

ykuijs committed:
Merge branch 'Dev' into MSFT_IntuneDeviceFeaturesConfigurationPolicyIOS-(new-module)

ykuijs commented on Modules/Microsoft365DSC/DSCResources/MSFT_IntuneDeviceFeaturesConfigurationPolicyIOS/MSFT_IntuneDeviceFeaturesConfigurationPolicyIOS.psm1:
```diff
@@ -0,0 +1,1167 @@
+        $AccessTokens
+
+    )
+    Write-Verbose -Message "Getting configuration of the Intune Device Features Configuration Policy for iOS with Id {$Id} and DisplayName {$DisplayName}"
```
This message says "Getting", however it is the Set method.

Please add this Get message to the Get method and change it here to Setting.

dannyKBjj replied:
done

ykuijs commented on Modules/Microsoft365DSC/DSCResources/MSFT_IntuneDeviceFeaturesConfigurationPolicyIOS/MSFT_IntuneDeviceFeaturesConfigurationPolicyIOS.psm1:
```diff
@@ -0,0 +1,1167 @@
+        $AccessTokens
+
+    )
+    #Ensure the proper dependencies are installed in the current environment.
```
Please add a Testing message here

dannyKBjj replied:
done

ykuijs reviewed (changes requested):
Few small comments

dannyKBjj committed:
Added requested message changes.

Added requested message changes.

dannyKBjj committed:
Merge branch 'MSFT_IntuneDeviceFeaturesConfigurationPolicyIOS-(new-module)' of https://github.com/dannyKBjj/Microsoft365DSC into MSFT_IntuneDeviceFeaturesConfigurationPolicyIOS-(new-module)

ykuijs committed:
Update CHANGELOG.md with new entries

ykuijs committed:
Merge branch 'Dev' into MSFT_IntuneDeviceFeaturesConfigurationPolicyIOS-(new-module)

ykuijs reviewed (approved):
LGTM
